### DISCUSSION TITLE
Startup event dependencies

### DISCUSSION THREAD

**Discussion by lockshaw** [fastapi/fastapi#8208] [Questions]
First of all, great work on FastAPI! I've been really impressed so far.

I need to start a long-running, periodic task on the startup event of my application. This long-running task requires the global app configuration. In my path handlers, I am using FastAPI's dependency injection to access the config, making it very easy to mock the dependency in tests. However, as of now I see no way to use dependency injection to get the config in the startup event. For now I can handle this by modifying the environ through starlette.environ, but it would be nice if I could mock the config in there the same way as I am doing the rest of the time. Is there a way to do this, or am I going about this in completely the wrong way?

**Comment by dmontagu** [comment]
@lockshaw I would use a pattern where the config is loaded by function call, and isn't loaded until the call. So then you can access it in your startup event without it being set prior to the test run. (Maybe I'm misunderstanding your architecture though.)

Something like:

```python
class AppSettings(BaseSettings):
    setting1: str
    setting2: str

@lru_cache()
def get_app_settings():
    return AppSettings()

# -------

@app.on_event("startup")
def trigger_repeat_using_settings():
    settings = get_app_settings()
    trigger_repetitions(settings)
```
Let me know if that doesn't make sense, or if there is a problem this doesn't address.

**Comment by lockshaw** [comment]
@dmontagu That's my plan, and it's good to have some confirmation that there's nothing I'm missing. My main question was simply if there was a nice way I could just keep using dependency injection as I do in the rest of the tests, but I guess not. Thanks!

**Comment by dmontagu** [comment]
@lockshaw the `app.on_event` function is inherited directly from starlette, so I don't think it can do anything nice with dependencies. But yes, I think this is the standard approach.

**Comment by tiangolo** [comment]
Thanks for the help @dmontagu ! :cake: :tada: 

Thanks @lockshaw for reporting back and closing the issue :heavy_check_mark:

**Comment by adriangb** [comment]
It would be very nice indeed to be able to use dependency injection in startup events!

For example, say you have a database connection.
I think this would look like:

```python
class AppSettings(BaseSettings):
    dbhost: str

class DataBase:
    def __init__(self, settings: AppSettings = Depends()):
        self._connection = Connection(url=settings.dbhost)

@app.on_event("startup")
def startup(db: DataBase = Depends()):
    app.state.db = db
```

Now, in your tests, I think you could just do:

```python
app.dependency_overrides[DataBase] = FakeDataBase
```

And there's no need to futz around with singletons or `lru_cache` because the startup event only ever happens once.

**Comment by ghost** [comment]
Can you make it possible to use Dependencies in events?

**Comment by meseta** [comment]
It would be nice to have a way built into FastAPI. I have the same dependency override use case as @adriangb, to change timeouts and URLs for integration tests.

In the mean time, I'm using this slightly unclean workaround, which only works when the dependency doesn't take additional arguments/dependencies of its own
```py
@app.on_event("startup")
async def startup_event():
    settings = app.dependency_overrides.get(get_settings, get_settings)()
```

This way, when I use `app.dependency_overrides[get_settings] = fake_get_settings` in a test. This is inspired by FastAPI's own dependency resolution which at some point calls `dependency_overrides.get(original_call, original_call)` in the `solve_dependencies()` method.

I had a poke around to see if I could write my own method that makes use of `solve_dependencies()`, and it appears possible, but some extra work is needed to get it to correctly resolve nested dependencies. Here's a PoC, that works for simple dependencies like above, but would require modifications to deal with nested ones.

NOTE: DO NOT USE, proof of concept only. (to deter copy/pasting, I have deleted some syntax) maybe someone could take it further

```py
from fastapi.dependencies.models import Dependant
from fastapi.dependencies.utils import solve_dependencies

async def call_single_dependency(app: FastAPI, get_dependency: Callable) -> Any:
    fake_scope = {
        "asgi": {"version": "3.0", "spec_version": "2.0"},
        "type": "http",
        "query_string": "",
        "http_version": "1.1",
        "method": "GET"
        "path": "",
        "root_path": ""
        "client": None,
        "server": None
        "headers": [],
        "extensions": {},
    }
    fake_request = Request(scope=fake_scope)
    values, *_ = await solve_dependencies(request=fake_request, dependant=Dependant(dependencies=[Dependant(path="", name="result", call=get_dependency)]), dependency_overrides_provider=app)
    return values["result"]
```

**Comment by ntakouris** [comment]
bump

**Comment by tandav** [comment]
My app has startup event which uses db session.  
I created I simple context manager wrapper. A bit ugly but it works.

```py
def get_db():
    db = SessionLocal()
    yield db
    db.close()

@app.on_event("startup")
def my_startup_event(db = Depends(get_db)):
    with get_db() as db:
        # ERROR:    Application startup failed. Exiting. 
        db.add(something)
        db.commit() 
```


```py

import contextlib
get_db_wrapper = contextlib.contextmanager(get_db)

@app.on_event("startup")
def my_startup_event():
    with get_db_wrapper() as db:
        # works
        db.add(something)
        db.commit() 
```

**Comment by xiki808** [comment]
@tandav thanks worked for me that way

**Reply by Matbj** [reply]
@tandav @xiki808 it is actually not using the `Depends(get_db)` the `db` attribute of the `my_startup_event` method is never used but instead shadowed in the `with... as db:`

**Comment by JWDobken** [comment]
The workaround of @meseta works for me when the dependency doesn't take additional arguments/dependencies of its own! And I worked with the contextmanager to trickle down dependencies :)

These are the dependencies to get an async session:

```py
from functools import lru_cache

from fastapi import Depends
from pydantic_settings import BaseSettings
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker


class Settings(BaseSettings):
    database_url: str


@lru_cache()
def get_settings():
    return Settings()


async def get_async_engine(settings: Settings = Depends(get_settings)) -> AsyncEngine:
    yield create_async_engine(
        settings.database_url,
    )


async def get_async_session(async_engine: AsyncEngine = Depends(get_async_engine)) -> AsyncSession:
    async_session = sessionmaker(bind=async_engine, class_=AsyncSession, expire_on_commit=False)
    async with async_session() as session:
        yield session
```

The startup event that works also with [Dependency overrides for tests](https://fastapi.tiangolo.com/advanced/testing-dependencies/):

```py
@app.on_event("startup")
async def startup_event():
    from app.init_db import init_db

    settings = app.dependency_overrides.get(get_settings, get_settings)()
    get_async_engine_context = contextlib.asynccontextmanager(get_async_engine)
    get_async_session_context = contextlib.asynccontextmanager(get_async_session)
    async with get_async_engine_context(settings=settings) as async_engine:
        async with get_async_session_context(async_engine=async_engine) as session:
            result = await session.execute(...)
```

**Comment by TarasKuzyo** [comment]
@JWDobken I am using your solution slightly modified to my use case 

```
async def get_session() -> AsyncGenerator[AsyncSession, None]:
    async with async_session() as session:
        async with run_transaction(session):
            yield session

@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
    get_async_session_context = asynccontextmanager(get_session)
    async with get_async_session_context() as session:
        scheduler = SchedulerService(session)
        scheduler.start()
        yield
```

my concern is the `yield` location. I have to place it inside the async session context in order to get it working.
Is it the correct approach? How do you handle it in your case?

**Comment by ntakouris** [comment]
DI Definitely needs to be built in. I wonder if we can RFC this and decide
on something. Perhaps influenced by the newest asp.net core di concepts
<https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0>...
something that you can just register using the `app`....
`app.register(<Interface Type> , <Implementation>)`

And then you can just use `InjectDependency(<Interface Type>, <(Optional)
Implementation Type Preference> )` on the controller or similar.... Perhaps
even a fastapi function to instantiate non-decorated functions....

`injected_partial_fn = app.inject_deps(..., fn)`

On Wed, Jan 3, 2024 at 8:34 PM Taras ***@***.***> wrote:

> @JWDobken <https://github.com/JWDobken> I am using your solution slightly
> modified to my use case
>
> async def get_session() -> AsyncGenerator[AsyncSession, None]:
>     async with async_session() as session:
>         async with run_transaction(session):
>             yield session
>
> @asynccontextmanager
> async def lifespan(app: FastAPI) -> AsyncGenerator[None, None]:
>     get_async_session_context = asynccontextmanager(get_session)
>     async with get_async_session_context() as session:
>         scheduler = SchedulerService(session)
>         scheduler.start()
>         yield
>
> my concern is the yield location. I have to place it inside the async
> session context in order to get it working.
> Is it the correct approach? How do you handle it in your case?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/discussions/8208#discussioncomment-8005899>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJPKMUFY4W3OODAHG6UXX3YMWQEZAVCNFSM6AAAAABAWFRQFKVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DAMBVHA4TS>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Best Regards,
Theodoros Ntakouris
+30 698 92 90 272 (Greece)
https://ntakour.is <https://ntakour.is/?src=gmail>
